### DISCUSSION TITLE
Picking dotnet tools up from manifest?

### DISCUSSION THREAD

**Discussion by nalla** [cake-build/cake#3269] [Q&A] [answered]
Hi, currently the following setup will not work.

```powershell
mkdir foo
cd foo
git init
dotnet new tool-manifest
dotnet tool install cake.tool --version 1.1.0
dotnet tool install gitversion.tool --version 5.6.6
echo "var v = GitVersion();var t = Argument(""target"",""Foo"");Task(""Foo"").Does(()=>{Information(v.SemVer);});RunTarget(t);" > build.cake
git add .
git commit -m"foo"
dotnet cake
Error: One or more errors occurred. (GitVersion: Could not locate executable.)
        GitVersion: Could not locate executable.
```

With this setup, `dotnet-gitversion` is only available with the command `dotnet gitversion`. Any ideas how to solve this? I would like not to use the `dotnet-tool` module because then there are two places inside the repository where tools are defined. Inside the tool manifest and inside a cake file. But maybe its too much I want to achieve here..

**Comment by augustoproiete** [comment, marked as answer]
This is not a supported scenario at the moment. We're working towards adding .NET local tool support via https://github.com/cake-build/cake/pull/3212 for an upcoming version of Cake.

If you only use GitVersion inside of the Cake build script, you could remove GitVersion from the tool-manifest and use the [.NET Tool Provider](https://cakebuild.net/docs/writing-builds/preprocessor-directives/tool/dotnet-provider), keeping the GitVersion files only in one place.

```cakescript
#tool "dotnet:?package=GitVersion.Tool&version=5.6.6"
```

Notice that as of [Cake v1.1](https://cakebuild.net/blog/2021/03/cake-v1.1.0-released) support for the `dotnet` module is bundled with Cake, so no need for using an external module anymore, and no need for bootstrapping either.

Alternatively, if you're _only_ building on Windows, you also have the option of using [GitVersion.CommandLine](https://www.nuget.org/packages/GitVersion.CommandLine/)

```cakescript
#tool "nuget:?package=GitVersion.CommandLine&version=5.6.6"
```

**Reply by gitfool** [reply]
Yes, this is the scenario I had in mind for #3212; ideally running the tool directly from the NuGet cache.

**Reply by nalla** [reply]
jeah well currently from the looks of it and the temporary folders and the speed that I can observe, the module approach with
```
#tool "dotnet:?package=GitVersion.Tool&version=5.6.6"
```
is way slower than the tools manifest. It does not seem to use the cache.

**Comment by nalla** [comment]
FYI, here is a workaround for the time being. I installed a helper function inside my build script package and then I'm using that function instead of the built in one.

```c#
GitVersion GetGitVersion()
{
	StartProcess("dotnet", new ProcessSettings
	{
		RedirectStandardOutput = true,
		RedirectStandardError = true,
		Silent = true,
		Arguments = "gitversion"
	}, out IEnumerable<string> output);
	return System.Text.Json.JsonSerializer.Deserialize<GitVersion>(string.Join(string.Empty, output));
}
```
